pkg/autodiscovery/providers/kubecontainer.go

```diff
@@ -0,0 +1,315 @@
+			if !ok {
+				return
+			}
+
+			k.processRuntimeEvents(evBundle)
+
+		case <-health.C:
+
+		}
+	}
+}
+
+func (k *KubeContainerConfigProvider) processOrchestratorEvents(evBundle workloadmeta.EventBundle) {
+	close(evBundle.Ch)
+
+	for _, event := range evBundle.Events {
+		switch event.Type {
+		case workloadmeta.EventTypeSet:
+			pod := event.Entity.(*workloadmeta.KubernetesPod)
+			k.addPod(pod)
+		case workloadmeta.EventTypeUnset:
+			pod := event.Entity.(*workloadmeta.KubernetesPod)
+			k.deletePod(pod)
+		default:
+			log.Errorf("cannot handle event of type %d", event.Type)
+		}
+	}
+}
+
+func (k *KubeContainerConfigProvider) processRuntimeEvents(evBundle workloadmeta.EventBundle) {
+	close(evBundle.Ch)
+
+	for _, event := range evBundle.Events {
+		switch event.Type {
+		case workloadmeta.EventTypeSet:
+			container := event.Entity.(*workloadmeta.Container)
+			k.addContainer(container)
+		case workloadmeta.EventTypeUnset:
+			container := event.Entity.(*workloadmeta.Container)
+			k.deleteContainer(container)
+		default:
+			log.Errorf("cannot handle event of type %d", event.Type)
+		}
+	}
+}
+
+func (k *KubeContainerConfigProvider) addContainer(e *workloadmeta.Container) {
+	k.Lock()
+	defer k.Unlock()
+	// TODO do 5 second container delay thing here
+	containerID := e.ID
+	k.containerCache[containerID] = e
+	log.Debugf("Adding entity container %s with labels %v\n", containerID, e.EntityMeta.Labels)
+	k.upToDate = false
+}
+
+func (k *KubeContainerConfigProvider) addPod(e *workloadmeta.KubernetesPod) {
+	k.Lock()
+	defer k.Unlock()
+	id := e.GetID().ID
+	k.podCache[id] = e
+	log.Debugf("adding pod with ID %s\n", id)
+	k.upToDate = false
+}
+
+func (k *KubeContainerConfigProvider) deleteContainer(e *workloadmeta.Container) {
+	k.Lock()
+	defer k.Unlock()
+	// TODO do 5 second container delay thing here
+	containerID := e.ID
+	log.Debugf("deleting container %s\n", containerID)
+	delete(k.containerCache, containerID)
+	k.upToDate = false
+}
+
+func (k *KubeContainerConfigProvider) deletePod(e *workloadmeta.KubernetesPod) {
+	k.Lock()
+	defer k.Unlock()
+	delete(k.podCache, e.GetID().ID)
+	log.Debugf("deleting pod with ID %s\n", e.GetID().ID)
+	k.upToDate = false
+}
+
+func (k *KubeContainerConfigProvider) generateConfigs() ([]integration.Config, error) {
+	k.Lock()
+	defer k.Unlock()
+
+	adErrors := make(map[string]ErrorMsgSet)
+
+	var configs []integration.Config
+
+	log.Debugf("Generating Configs for %d containers\n", len(k.containerCache))
+	for containerID, container := range k.containerCache {
+		containerEntityName := containers.BuildEntityName(string(container.Runtime), containerID)
+		c, errors := utils.ExtractTemplatesFromContainerLabels(containerEntityName, container.Labels)
+
+		for _, err := range errors {
+			log.Errorf("Can't parse template for container %s: %s", containerID, err)
+			// TODO should add to 'adErrors'? Different from 'ContainerConfigProvider' but seems more correct
+		}
+
+		if util.CcaInAD() {
```

Comment:
I'm ambivalent on whether this provider needs a feature flag of its own (I don't think so, but you may find it useful to turn it off separately), but I do think that this should NOT be available without CCA in AD (so we'd get rid of these checks in the provider), since the old collectors are stable and work correctly right now.

Reply:
Sounds good to me, I'll just put this behind the `CcaInAd` check and it can roll out with CCA in AD.
I don't feel strongly about it needing a second feature flag specific to this provider.